contracts/Prover.sol

```diff
@@ -154,17 +149,103 @@ contract Prover is SimpleProver, AbstractProver {
+    function getProofType() external pure override returns (ProofType) {
+        return PROOF_TYPE;
+    }
+
+    function _setChainConfiguration(
+        ChainConfigurationKey memory chainConfigurationKey,
+        ChainConfiguration memory chainConfiguration
+    ) internal {
+        chainConfigurations[chainConfigurationKey.chainId][chainConfigurationKey.provingMechanism] = chainConfiguration;
+        // if (chainConfigurationKey.chainId == block.chainid) {
         l1BlockhashOracle = IL1Block(chainConfiguration.blockhashOracle);
+        // }
+    }
+
+    function getProvenState(uint256 chainId, ProvingMechanism provingMechanism)
+        internal
+        view
+        returns (
+            ChainConfiguration memory chainConfiguration,
+            BlockProofKey memory blockProofKey,
+            BlockProof memory blockProof
+        )
+    {
+        if (provingMechanism == ProvingMechanism.Bedrock) {
+            chainConfiguration = chainConfigurations[chainId][ProvingMechanism.Bedrock];
+            {
+                if (chainConfiguration.settlementChainId != block.chainid) {
+                    blockProof = provenStates[chainConfiguration.settlementChainId][SettlementType.Finalized];
+                    blockProofKey = BlockProofKey({chainId: chainId, settlementType: SettlementType.Finalized});
+                } else {
+                    blockProof = provenStates[chainConfiguration.settlementChainId][SettlementType.Confirmed];
+                    blockProofKey = BlockProofKey({chainId: chainId, settlementType: SettlementType.Confirmed});
+                }
+            }
+        }
+        return (chainConfiguration, blockProofKey, blockProof);
+    }
+
+    // To see block information available on chain see
+    // https://docs.soliditylang.org/en/latest/units-and-global-variables.html#block-and-transaction-properties
+    /**
+     * @notice validates input block state against the last 256 blocks on chain
+     * @param rlpEncodedBlockData properly encoded block data
+     * @dev inputting the correct block's data encoded as expected will result in its hash matching
+     * the blockhash found on the last 256 blocks on chain. This means that the world state root found
+     * in that block represents a valid state.
+     */
+    function proveSelfState(bytes calldata rlpEncodedBlockData) public {
+        if (!chainConfigurations[block.chainid][ProvingMechanism.Self].exists) {
+            revert InvalidDestinationProvingMechanism(block.chainid, ProvingMechanism.Self);
+        }
+        BlockProof memory blockProof = BlockProof({
+            blockNumber: bytesToUint(RLPReader.readBytes(RLPReader.readList(rlpEncodedBlockData)[8])),
+            blockHash: keccak256(rlpEncodedBlockData),
+            stateRoot: bytes32(RLPReader.readBytes(RLPReader.readList(rlpEncodedBlockData)[3]))
+        });
+        require(
+            blockhash(blockProof.blockNumber) == blockProof.blockHash,
```

Comment:
is the purpose of this to confirm at the blockash in the rlpEncodedData is correct? or that the block is in the last 256 blocks? or both? The error should reflect this.

Reply:
if its just to make sure the block is recent enough, you could just check whether the blockhash is 0 -  this method returns 0 when the blocknumber is not within the past 256.